### PR TITLE
superpmi.py script improvements

### PR DESCRIPTION
**Pull request by BruceForstall**
1. Simplify and regularize superpmi and mcs tool finding: they are both looked
for first in Core_Root, then in the built product directory, then on the PATH.
2. Improve JIT-EE version determination: first, search for the GUID by parsing
corinfo.h, if found. Otherwise, use "mcs -printJITEEVersion" as before.
3. Improve JIT dll finding: search first in Core_Root, then in the built
product directory.
4. When downloading coredistools, create the Core_Root if it doesn't exist.

With these changes, you can build the "clr.runtime" subset then immediately run
a replay or asmdiffs, without first generating a Core_Root directory. The script
will automatically create the Core_Root directory simply to put coredistools in
it (and also use it as the "current directory" when running superpmi commands).

### CODE REVIEW

**Comment by BruceForstall** [comment]
cc @dotnet/jit-contrib

File: src/coreclr/scripts/superpmi.py [hunk 1]
```diff
@@ -1690,16 +1735,94 @@ def determine_superpmi_tool_path(coreclr_args):
...
+               0x4176,
+               0x43a7,
+               {0x8c, 0x2b, 0xa0, 0x5b, 0x55, 0x1d, 0x4f, 0x49}
+           };
+
+           We want the string between the /* */ comments.
+        2. Find the mcs tool and run "mcs -printJITEEVersion".
+        3. Otherwise, just use "unknown-jit-ee-version", which will probably cause downstream failures.
+
+        NOTE: When using mcs, we need to run the tool. So we need a version that will run. If a user specifies
```
Comment (kunalspathak) [inline src/coreclr/scripts/superpmi.py:1787]: We discussed lot of things today, so it is easy for me to forget. But remind me, even we will figure out guid in `superpmi.py` script itself, but in #42053 , I will still need `coreclr` x64 so I can run `mcs.exe` (eventually via `superpmi.py`) to merge/dedup .mcs files. Yes?

File: src/coreclr/scripts/superpmi.py [hunk 2]
```diff
@@ -1690,16 +1735,94 @@ def determine_superpmi_tool_path(coreclr_args):
...
+    if os.path.isfile(corinfo_h_path):
+        # The string is near the beginning of the somewhat large file, so just read a line at a time when searching.
+        with open(corinfo_h_path, 'r') as file_handle:
+            for line in file_handle:
+                match_obj = re.search(r'JITEEVersionIdentifier *= *{ */\* *([^ ]*) *\*/', line)
+                if match_obj is not None:
+                    corinfo_h_jit_ee_version = match_obj.group(1)
+                    print("Using JIT/EE Version from corinfo.h: {}".format(corinfo_h_jit_ee_version))
+                    return corinfo_h_jit_ee_version
+            print("Warning: couldn't find JITEEVersionIdentifier in {}; is the file corrupt?".format(corinfo_h_path))
```
Comment (kunalspathak) [inline src/coreclr/scripts/superpmi.py:1810]: nit: put this under `if verbose`?

**Review by kunalspathak — APPROVED** [review]
LGTM with a question.

File: src/coreclr/scripts/superpmi.py [hunk 3]
```diff
@@ -2565,7 +2659,7 @@ def verify_replay_common_args():
 
         # Make a more intelligent decision about the arch and build type
         # based on the path of the jit passed
-        if standard_location and not coreclr_args.build_type in coreclr_args.jit_path:
+        if jit_in_product_location and not coreclr_args.build_type in coreclr_args.jit_path:
```
Comment (am11) [inline src/coreclr/scripts/superpmi.py:2662]: a nit: one operator versus two
```suggestion
        if jit_in_product_location and coreclr_args.build_type not in coreclr_args.jit_path:
```

File: src/coreclr/scripts/superpmi.py [hunk 4]
```diff
@@ -2671,13 +2765,13 @@ def verify_replay_common_args():
 
         # Make a more intelligent decision about the arch and build type
         # based on the path of the jit passed
-        if standard_location and not coreclr_args.build_type in coreclr_args.base_jit_path:
+        if jit_in_product_location and not coreclr_args.build_type in coreclr_args.base_jit_path:
```
Comment (am11) [inline src/coreclr/scripts/superpmi.py:2768]:
```suggestion
        if jit_in_product_location and coreclr_args.build_type not in coreclr_args.base_jit_path:
```

File: src/coreclr/scripts/superpmi.py [hunk 5]
```diff
@@ -2671,13 +2765,13 @@ def verify_replay_common_args():
...
             coreclr_args.verify(determined_build_type,
                                 "build_type",
                                 coreclr_args.check_build_type,
                                 "Invalid build_type")
 
-        if standard_location and not coreclr_args.arch in coreclr_args.base_jit_path:
+        if jit_in_product_location and not coreclr_args.arch in coreclr_args.base_jit_path:
```
Comment (am11) [inline src/coreclr/scripts/superpmi.py:2774]:
```suggestion
        if jit_in_product_location and coreclr_args.arch not in coreclr_args.base_jit_path:
```

File: src/coreclr/scripts/superpmi.py (hunk 1 above)
Comment (BruceForstall) [reply src/coreclr/scripts/superpmi.py:1787]: > I will still need coreclr x64 so I can run mcs.exe (eventually via superpmi.py) to merge/dedup .mcs files. Yes?

Yes.

File: src/coreclr/scripts/superpmi.py (hunk 2 above)
Comment (BruceForstall) [reply src/coreclr/scripts/superpmi.py:1810]: Well, it's a pretty bad situation, so it probably warrants always being displayed.

I'm planning to revamp the whole output situation, creating log files, verbosity levels, etc.

**Pull request merged**